### ISSUE TITLE
Can darknet detector demo use multiple GPUs?

### ISSUE THREAD

**Issue by snobu** [pjreddie/darknet#777]
Can `darknet detector demo` use multiple GPUs?

`-gpus` doesn't seem to kick off a 2nd process on the 2nd GPU.
```
$ ./darknet detector demo cfg/coco.data cfg/yolov3.cfg yolov3.weights Delhi.mkv -gpus 0,1
```
...
```
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 390.46                 Driver Version: 390.46                    |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|===============================+======================+======================|
|   0  Tesla M60           Off  | 00003DB5:00:00.0 Off |                  Off |
| N/A   54C    P0   135W / 150W |   1450MiB /  8129MiB |     89%      Default |
+-------------------------------+----------------------+----------------------+
|   1  Tesla M60           Off  | 000057C7:00:00.0 Off |                  Off |
| N/A   40C    P8    14W / 150W |     11MiB /  8129MiB |      0%      Default |
+-------------------------------+----------------------+----------------------+

+-----------------------------------------------------------------------------+
| Processes:                                                       GPU Memory |
|  GPU       PID   Type   Process name                             Usage      |
|=============================================================================|
|    0     15257      C   ./darknet                                   1439MiB |
+-----------------------------------------------------------------------------+
```

Is multi-GPU support just for training?

**Comment by TheMikeyR** [comment]
multi-gpu is only supported in the training step.

**Comment by snobu** [comment]
Thanks!

**Comment by snobu** [comment]
Out of pure curiosity, is this feature a difficult proposition technically or your model simply can't benefit from splitting the work across GPUs?

**Comment by Benz-Tracxpoint** [comment]
Hi,
Thanks for the quick response!

> > ./darknet detector train cfg/my.data cfg/my.cfg yolov3.weights labels=cfg/my.names -gpus 0 -dont_show -clear 1

> As desribed here you should use `darknet53.conv.74` pre-trained file instead of `yolov3.weights`: https://github.com/AlexeyAB/darknet#how-to-train-to-detect-your-custom-objects

> `./darknet detector train data/obj.data yolo-obj.cfg darknet53.conv.74`

> > ./darknet detector train cfg/my.data cfg/my.cfg yolov3.weights labels=cfg/my.names -gpus 0 -dont_show -clear 1

> > What does it mean `labels=cfg/my.names` ?

For my next run i will use the following:
`./darknet detector train cfg/my.data cfg/my.cfg darknet53.conv.74  -gpus 0 -dont_show -clear 1`
Am I training with multiple GPUs in the correct way? waiting for 1000 steps weights and then call the command:
`./darknet detector train cfg/my.data cfg/my.cfg my_1000.weights -gpus 0,1,2,3 -dont_show -clear 1`
?

> > When trying to predict using the weights i have created I am getting a few weird issues
> > first: I have gotten this weird error message:
> > Error: in the file data/coco.names number of names 80 that isn't equal to classes=89 in the file cfg/my.cfg

> > As described here: https://github.com/AlexeyAB/darknet#how-to-train-to-detect-your-custom-objects

> > Note: After training use such command for detection: darknet.exe detector test data/obj.data yolo-obj.cfg yolo-obj_8000.weights

> You should use test command like this:
> `./darknet detector test cfg/my.data cfg/my.cfg my_10000.weights dog.jpg`

Tried it with the weights files i have - still no detection, i will try it after running the training again with the advices weights file today.

> * Do you use this repository https://github.com/AlexeyAB/darknet or this https://github.com/pjreddie/darknet ?
> * What is the date of your Darknet code, when you clone it?

This is the last log i have checked out:
>	commit 21a4ec9390b61c0baa7ef72e72e59fa143daba4c
>	Author: AlexeyAB <alexeyab84@gmail.com>
>	Date:   Mon Nov 26 11:11:56 2018 +0300
>
>		Saving loss-chart for each 100 iterations automatically

> * What parameters do you use in the Makefile?
Not sure, the default ones.
> * And what GPU do you use?
I am training it on 4 x Tesla V100 GPUs

> * What versions of CUDA and cuDNN do you use? Does cuDNN version correspond to the CUDA version? You shouldn't use cuDNN 7.3 for CUDA 9.0 with CUDA 10.0
CUDNN 7.4.0, CUDA: V10.0.130

> * Try to check mAP for `my_10000.weights` as described here: https://github.com/AlexeyAB/darknet#when-should-i-stop-training

> `./darknet detector map cfg/my.data cfg/my.cfg my_10000.weights`
and
> `./darknet detector map cfg/my.data cfg/my.cfg my_20000.weights`

So if I have 89 classes I am looking for 89*2000 iterations in total(178,000) right? I will try using this command during the next training session

**Comment by slmatrix** [comment]
> multi-gpu is only supported in the training step.

how is it enables? Only one card is baking for me.

**Comment by slmatrix** [comment]
train with multiple gpus:
**> ./darknet detector train <data> <config> [optional_weights] _-gpus 0,1,..,..._**

**Comment by folkien** [comment]
> multi-gpu is only supported in the training step.

Thanks for reply. So i have another question - is it possible to run multiple YOLO instances spread across multiple GPU's? @pjreddie @AlexeyAB

**Comment by Snickbrack** [comment]
I‘m also interested in this @folkien

**Comment by folkien** [comment]
@Snickbrack I've tested this in python and it's possible tu run multiple YOLO's on multiple GPU and it works well.

**Comment by uwing85213** [comment]
@folkien 
Hey, bro!
I tested to run 2 Yolo models on one GPU recently, but the result left much to be desired.
So I want to run 2 models on each 2 GPU respectively. Unfortunately, it is too difficult to create for me.
Could you share your project for me?
Thanks!

**Comment by folkien** [comment]
You have to change gpu before initialization of YOLO. Using darknet.py (https://github.com/pjreddie/darknet/blob/master/python/darknet.py) you can code something like that : 

```python
        # Using GPU 0
        darknet.set_gpu(0)
        self.net, self.classes, self.colors = darknet.load_network(
            cfgPath, metaPath, weightPath)

        # Using GPU 1
        darknet.set_gpu(1)
        self.net, self.classes, self.colors = darknet.load_network(
            cfgPath, metaPath, weightPath)
```

**Comment by uwing85213** [comment]
@folkien 
Thanks so much!

**Comment by reddevil1310** [comment]
> @folkien
> Hey, bro!
> I tested to run 2 Yolo models on one GPU recently, but the result left much to be desired.
> So I want to run 2 models on each 2 GPU respectively. Unfortunately, it is too difficult to create for me.
> Could you share your project for me?
> Thanks!

@uwing85213  How run 2 Yolo models on one GPU at the same time?
Thanks.

**Comment by uwing85213** [comment]
@reddevil1310 

You don’t need to specify a GPU, just load two models.

I referred to this project.
https://github.com/AlexeyAB/darknet

**Comment by folkien** [comment]
> > @folkien
> > Hey, bro!
> > I tested to run 2 Yolo models on one GPU recently, but the result left much to be desired.
> > So I want to run 2 models on each 2 GPU respectively. Unfortunately, it is too difficult to create for me.
> > Could you share your project for me?
> > Thanks!
> 
> @uwing85213 How run 2 Yolo models on one GPU at the same time?
> Thanks.

You can run as many yolo models as you want on single GPU, there is only 1 limit - GPU memory! For example i'm using RTX2060 and i can fit 3-4models of YOLOv4 per GPU.

**Comment by reddevil1310** [comment]
> @ reddevil1310
> 
> Bạn không cần chỉ định GPU, chỉ cần tải hai mô hình.
> 
> Tôi đã tham khảo dự án này.
> https://github.com/AlexeyAB/darknet

Thank you!

**Comment by folkien** [comment]
@Snickbrack @reddevil1310 @uwing85213 @AlexeyAB My new interesiting observation. I have sucessfuly run YOLOv4 models simultanously on mutiple same model GPU's  (for example two RTX2060). Also tested on two RTX3070. But there is a problem when running in parallel simultanously YOLOv4 detectors with mixed GPU's model - connected to motherboard 1x RTX2060 and 1xRTX3070. I don't know problem source, is it driver or is it CUDA or is it darknet?

Models are loaded sucessfuly but doesn't give any detections during run. If i use mixed hardware(RTX2060 and RTX3070) i can sucesfully run YOLOv4 on single GPU (doesn't matter if it's RTX2060 or RTX3070) but problem occurs only if i would like to run code on both GPU's in parallel. 

I didn't investigated it further, but only changed hardware to have 2xRTX2060 in first PC and 2xRTX3070 in second PC.

**Comment by AlexeyAB** [comment]
@folkien 

> Models are loaded sucessfuly but doesn't give any detections during run. 

1. Does Yolo stops? Or does Yolo give the result without bboxing? What do you see in the terimnal? 
2. Check the CC (CUDA Compute Capability), try to compile Darknet with CC 3.5 or 5.2
2. Does it work well with CUDNN_HALF=0 ?
3. Maybe there is an issue with CUDA + cuDNN + nVidia Driver

**Comment by folkien** [comment]
> @folkien
> 
> > Models are loaded sucessfuly but doesn't give any detections during run.
> 
>     1. Does Yolo stops? Or does Yolo give the result without bboxing? What do you see in the terimnal?
> 
>     2. Check the CC (CUDA Compute Capability), try to compile Darknet with CC 3.5 or 5.2
> 
>     3. Does it work well with CUDNN_HALF=0 ?
> 
>     4. Maybe there is an issue with CUDA + cuDNN + nVidia Driver

Ok, in free time i will check this. First i was thinking that there is a driver problem, but i had `nvidia-driver-465` which supports both rtx 2060 and rtx 3070. Error confirmed on both ubuntu 20.04 and 18.04 with `nvidia-driver-465` and CUDA 10.2 and darknet master  4f794aa15265d2a068d9f452e5efd9d7a3574524.

**Comment by gh-develop** [comment]
@uwing85213
Hello, sir. I saw what you wrote.
'I tested to run 2 Yolo models on one GPU recently, but the result left much to be desired.'

I referred to this project.
https://github.com/AlexeyAB/darknet

I tried modifying darknet_video.py, but it didn't work properly.
Can you give me some advice?

**Comment by aristosamar** [comment]
I had similar problem when trying to use GTX1070 and RTX4060 at this same time. Training model works when using one of GPUs but not both at the same time: (darknet detector train ... -gpus 0 or -gpus 1 but not -gpus 0,1
